### ISSUE TITLE
[BUG] MediaElement doesn't work on non-root pages - iOS

### ISSUE THREAD

**Issue by RomanShaparev** [CommunityToolkit/Maui#1778]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Did you read the "Reporting a bug" section on Contributing file?

- [X] I have read the "Reporting a bug" section on Contributing file: https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

### Current Behavior

**iOS**: When using NavigationPage and navigating to a page containing a MediaElement, an error occurs: 
`ObjCRuntime.ObjCException: Objective-C exception thrown.  Name: CALayerInvalid Reason: layer <CALayer: 0x6000007e4e60> is a part of cycle in its layer tree`

### Expected Behavior

Should work fine when Push'ing to a page containing a MediaElement.

### Steps To Reproduce

1. Launch the iOS version from the reproduction repository.
2. Click **Push MediaElementPage** button

### Link to public reproduction project repository

https://github.com/RomanShaparev/MediaElementCrash

### Environment

```markdown
- .NET MAUI CommunityToolkit.Maui.MediaElement: 3.1.0
- OS: iOS 17.0
- .NET MAUI: 8.0.14
```

**Comment by ne0rrmatrix** [comment]
Tested against current main and cannot reproduce.

**Comment by AntAttakk** [comment]
@ne0rrmatrix could you elaborate on what you mean by not adding media controls? 

It would be great if there was a workaround for this issue. We can't use the latest version of the MediaElement at all - the iOS app just crashes with the Objective-C exception mentioned above as soon as we are trying to show it.

We already have `ShouldShowPlaybackControls="False"`. Is this what you mean by not adding media controls? It does not help for us, I am afraid.

**Comment by ne0rrmatrix** [comment]
Comment out line 43 and build from source and your issue should be resolved. This is a temporary solution as it will prevent playback controls from being displayed at all. 

[https://github.com/CommunityToolkit/Maui/blob/main/src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs](https://github.com/CommunityToolkit/Maui/blob/main/src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs)
![image](https://github.com/CommunityToolkit/Maui/assets/4167863/a9bbe55c-a176-4ee9-a34b-1e3fd133b975)

I built a copy myself to test with the above mentioned highlighted in yellow commented out. It played the video. Just with no video controls. I hope this helps you. I am working on finding a fix for this.

**Comment by rafalka** [comment]
I think that entire idea with using `WindowStateManager.Default.GetCurrentUIViewController()` here is wrong.
It works only when use `MediaElement` on a root screen. 
The problem is that if you going to create a popup that opens a new window, this window is not added to a stack when you attempt to create a `MauiMediaElement` so view controller returned by `GetCurrentUIViewController()` is not the right one.

**Comment by ne0rrmatrix** [comment]
That line adds the controls for the video player. We have been looking for a better solution and have yet to find one. Until recently the code above worked fine. Also there is an issue with AVPlayer controls in general. We have had ongoing issues since IOS 16.x. You are not wrong with you answer. I just don't have an answer. 

As far as I am aware the code does not create a popup at all. There is nothing in the code to do that for IOS. All it does is create a view controller and add avplayer to it.

**Comment by rafalka** [comment]
@ne0rrmatrix: Yes, MAUI itself doesn't open new windows but plugins like [Mopups](https://github.com/LuckyDucko/Mopups) does.

I think that this can be solved by overriding ` WillMoveToWindow(UIWindow? window)` and getting a ViewController directly from window. 

However as I checked this code:
```CSharp
	// Add the View from the AVPlayerViewController to the parent ViewController
	if (viewController is not ShellFlyoutRenderer && viewController is not PageViewController && viewController is not UICollectionViewController)
	{
		viewController.AddChildViewController(playerViewController);
	}
	viewController.View.AddSubview(playerViewController.View);
```
run on actual window VC gives an opposite effect: adding AddChildViewController() makes controls invisible and `viewController.View.AddSubview()` make player window invisible at all.

**Comment by rafalka** [comment]
While waiting for a real fix you can try this workaround. 
Just replace `MediaElement` handler with this one:
```CSharp
public class MediaElementHandlerFix : MediaElementHandler
{
    protected override MauiMediaElement CreatePlatformView()
    {
        if (MauiContext is null)
        {
            throw new InvalidOperationException($"{nameof(MauiContext)} cannot be null");
        }

        mediaManager ??= new(MauiContext,
            VirtualView,
            Dispatcher.GetForCurrentThread() ?? throw new InvalidOperationException($"{nameof(IDispatcher)} cannot be null"));

        var (_, playerViewController) = mediaManager.CreatePlatformView();
        var page = VirtualView.FindParent<Page>();
        var parentViewController = (page?.Handler as PageHandler)?.ViewController;

        return new MauiMediaElementFix(playerViewController, parentViewController);
    }
}

internal class MauiMediaElementFix : MauiMediaElement
{
    private readonly AVPlayerViewController _playerViewController;

    public MauiMediaElementFix(AVPlayerViewController playerViewController, UIViewController? parentViewController) 
#pragma warning disable CS0618 // Type or member is obsolete - required for workarounding bug https://github.com/CommunityToolkit/Maui/issues/1778
        : base(playerViewController, new DummyUiViewController())
#pragma warning restore CS0618 // Type or member is obsolete
    {
        _playerViewController = playerViewController;
        if (parentViewController != null)
        {
            if (parentViewController.View != null)
            {
                var insets = parentViewController.View.SafeAreaInsets;
                _playerViewController.AdditionalSafeAreaInsets =
                    new UIEdgeInsets(insets.Top * -1, insets.Left, insets.Bottom * -1, insets.Right);
            }

            parentViewController.AddChildViewController(_playerViewController);
        }
    }

    public override void WillMoveToWindow(UIWindow? window)
    {
        if (window is null)
        {
            _playerViewController.Player?.Pause();
            _playerViewController.RemoveFromParentViewController();
        }
        
        base.WillMoveToWindow(window);
    }
}
// We need that dummy VC for disabling setting parent VC logic in MauiMediaElement constructor
internal class DummyUiViewController : UIViewController
{
    public override UIView? View
    {
        get => null;
        set { /*ignore*/ }
    }
}
```
it depends on an extension method:
```CSharp
    public static T? FindParent<T>(this VisualElement? child) where T : VisualElement
    {
        while (true)
        {
            if (child == null) return null;
            if (child.Parent is T parent) return parent;
            child = child.Parent as VisualElement;
        }
    }
```

**Comment by ne0rrmatrix** [comment]
@rafalka I tried your code. It looks like it actually fixes the issue. I tested against a shell and navigation page so far. Using the code you provided I was able to get Media Element to display the video, have the controls work, and it does not have any impact on full screen. Thank you. I will have a PR out in about 20 min for people to test. I will be giving you full credit for the contribution you provided.
